lib/vdc/DerivedVar.cpp

```diff
@@ -2899,3 +2899,390 @@ void DerivedCoordVarStandardOceanSCoordinate::compute_g2(
+		SetErrMsg("Invalid file descriptor : %d", fd);
+		return(-1);
+	}
+
+	_fileTable.RemoveEntry(fd);
+	delete f;
+
+	return(0);
+}
+
+int DerivedCoordVarStandardAHSPC::ReadRegion(
+	int fd,
+    const vector <size_t> &min, const vector <size_t> &max, float *region
+) {
+
+	DC::FileTable::FileObject *f = _fileTable.GetEntry(fd);
+
+	vector <size_t> dims, dummy;
+	int rc = GetDimLensAtLevel(f->GetLevel(), dims, dummy);
+	if (rc<0) return(-1);
+
+	string aVar = _aVar.empty() ? _apVar : _aVar;
+	vector <float> a(dims[2]);
+	vector <size_t> myMin = {min[2]};
+	vector <size_t> myMax = {max[2]};
+	rc = _getVar(
+		_dc, f->GetTS(), aVar, f->GetLevel(), f->GetLOD(),
+		myMin, myMax, a.data()
+	);
+	if (rc<0) return(rc);
+
+	vector <float> b(dims[2]);
+	myMin = {min[2]};
+	myMax = {max[2]};
+	rc = _getVar(
+		_dc, f->GetTS(), _bVar, f->GetLevel(), f->GetLOD(),
+		myMin, myMax, b.data()
+	);
+	if (rc<0) return(rc);
+
+	vector <float> ps(dims[0]*dims[1]);
+	myMin = {min[0], min[1]};
+	myMax = {max[0], max[1]};
+	rc = _getVar(
+		_dc, f->GetTS(), _psVar, f->GetLevel(), f->GetLOD(),
+		myMin, myMax, ps.data()
+	);
+	if (rc<0) return(rc);
+
+
+	if (! _aVar.empty()) {
+		float p0;
+		myMin = {};
+		myMax = {};
+		rc = _getVar(
+			_dc, f->GetTS(), _p0Var, f->GetLevel(), f->GetLOD(),
+			myMin, myMax, &p0
+		);
+
+		compute_a(
+			min, max, a.data(), b.data(), ps.data(), p0, region
+		);
+	}
+	else {
+		compute_ap(
+			min, max, a.data(), b.data(), ps.data(), region
+		);
+	}
+
+	return(0);
+}
+
+bool DerivedCoordVarStandardAHSPC::VariableExists(
+	size_t ts,
+	int reflevel,
+	int lod
+) const {
+
+	if (! _aVar.empty()) {
+		return(
+			_dc->VariableExists(ts, _aVar, reflevel, lod) &&
+			_dc->VariableExists(ts, _bVar, reflevel, lod) &&
+			_dc->VariableExists(ts, _psVar, reflevel, lod) 
+		);
+	}
+	else {
+		return(
+			_dc->VariableExists(ts, _apVar, reflevel, lod) &&
+			_dc->VariableExists(ts, _bVar, reflevel, lod) &&
+			_dc->VariableExists(ts, _psVar, reflevel, lod) &&
+			_dc->VariableExists(ts, _p0Var, reflevel, lod) 
+		);
+	}
+}
+
+bool DerivedCoordVarStandardAHSPC::ValidFormula(string formula) {
+
+	return(
+		DerivedCFVertCoordVar::ValidFormula(
+			vector <string> {"a", "b", "p0", "ps"}, formula
+		) ||
+		DerivedCFVertCoordVar::ValidFormula(
+			vector <string> {"ap", "b", "ps"}, formula
+		)
+	);
+}
+
+void DerivedCoordVarStandardAHSPC::compute_a(
+	const vector <size_t> &min, const vector <size_t> &max,
+	const float *a, const float *b, const float *ps, float p0,
+	float *region
+) const {
+
+	vector <size_t> rDims;
+	for (int i=0; i<3; i++) {
+		rDims.push_back(max[i]-min[i]+1);
+	}
+
+    for (size_t k=0; k<max[2]-min[2]+1; k++) {
+    for (size_t j=0; j<max[1]-min[1]+1; j++) {
+    for (size_t i=0; i<max[0]-min[0]+1; i++) {
+
+		// We are deriving coordinate values from data values, so missing
+		// values may be present
+		//
+		float l_ps = ps[j*rDims[0]+i];
+		if (l_ps == _psVarMV) l_ps = 0;
+
+		float pressure = (a[k]*p0) + (b[k]*l_ps);
```

Comment:
`compute_a` and `compute_ap` are identical other than the multiplication by `p0` here. You could just have one function with `p0` as an optional parameter initialized to `1`.

Reply:
Doh! Good catch. Fixed.